design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,217 @@
+An example of an attributes entry in the BIOS registry is:
+
+    {
+        "AttributeName": "SriovGlobalEnable",
+        "CurrentValue": null,
+        "DisplayName": "SR-IOV Global Enable",
+        "DisplayOrder": 2331,
+        "HelpText": "Enables or disables the BIOS configuration of Single Root
+                     I/O Virtualization (SR-IOV) devices. Enable this feature
+                     if booting to a virtualization operating system that
+                     recognize SR-IOV devices. ",
+        "Hidden": false,
+        "Immutable": false,
+        "MenuPath": "./IntegratedDevicesRef",
+        "ReadOnly": false,
+        "ResetRequired": true,
+        "Type": "Enumeration",
+        "Value": [
+          {
+            "ValueDisplayName": "Enabled",
+            "ValueName": "Enabled"
+          },
+          {
+            "ValueDisplayName": "Disabled",
+            "ValueName": "Disabled"
+          }
+        ],
+        "WarningText": null,
+        "WriteOnly": false
+    }
+
+The cached BIOS registry from Ironic should be made available to the user
+via an API. The BIOSRegistry can be quite large (up to 500 Kb uncompressed
+on some vendors) and is only useful on the reference host, not all of the hosts
+that the BIOS config will be pushed to.
+
+Using the BIOS settings and BIOS registry, the user will be able create a JSON
+list of bulk BIOS settings. An additional CRD will be defined describing these
+bulk settings. This JSON list can be set on multiple machines to load and
+apply, effectively cloning all the settings from another machine. The BMO will
+send the data to Ironic via the clean-steps API without validating or
+interpreting it. Ironic will convert the JSON to a set of clean steps and run
+cleaning to configure the BIOS settings.
+
+### Implementation Details/Notes/Constraints
+
+One of the primary drivers of this design was the desire to not have the
+BMO validate the user inputted bulk BIOS settings as it requires vendor
+specific checks that are difficult to maintain. This proposal does not suggest
+validity checks in the BMO or Ironic either - the input JSON list is converted
+to clean steps as-is.
+
+In order to avoid applying attributes that are ReadOnly when setting BIOS
+configuration, which will have no effect and will increase the time for BIOS
+configuration it is recommended that the user remove ReadOnly fields, as
+identified in the BIOS Registry, when building the JSON input. It’s not
+expected that the entire set read from Ironic will be passed down as:
+
+1. many of the values read from Ironic are ReadOnly
+2. setting ReadOnly values to their current settings increases the size of the
+   input and may potentially affect the Redfish request to the BMC
+
+The BIOS Settings are changed during the node cleaning step which requires a
+reboot. The change to BIOS settings will not be a trigger for reprovisioning a
+host, the BIOS settings will only be applied when a new image is going to be
```

Comment:
TBH I thought this was clearer before you rewrote it. How it will works is:
* if the Host is in the Ready state (i.e. provisioning has not started)
* if the settings in the status differ from the ones in the spec
* then the Host will re-enter the Preparing state
* in the Preparing state the settings get copied from the status to the spec
* in the Preparing state, manual cleaning takes place with the new settings
* after Preparing, the Host returns to Ready
* when provisioning, the node is already configured with the latest setttings

The previous statement captured that at a high level, including the fact that it only happens when the Host is Ready:

> Manual clean occurs in the Preparing state of the BaremetalHost, the
> Host will re-enter this state from Ready/Available state whenever its config
> differs from the last one it stored.

Whereas this one is actually incorrect - the settings are not applied when a new image is going to be written, they're applied before that. They'll be applied even if the host _never_ has an image written.

Reply:
I understand that the timing in ironic depends on specific states that may not line up with what we have in our API. I think we should limit ourselves to making changes to the host "during provisioning" which I interpret as "taking all of the settings given in the metal3 API and applying them to the host through ironic". So, I don't think we want Ironic to make BIOS settings until we know we're going to be asked to write an image to the host.

Is that possible?

Reply:
The BIOS settings are written to Ironic as a set of clean steps during manual cleaning, e.g. https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L1318

Reply:
The manual cleaning flow has already been implemented (that's what the Preparing state is for - see metal3-io/baremetal-operator#763) and at least as far as we know works correctly with the Ironic states, so implementation is just a matter of adding new steps to the list we send. The current implementation took 17 months to land, so we're not going to start again from scratch :)